src/http2.rs

```diff
@@ -0,0 +1,563 @@
+                                       "Not in a writable state")),
+            Some(ref mut req) => req.body.flush(),
+        }
+    }
+}
+
+impl<S> Read for Http2Message<S> where S: CloneableStream {
+    #[inline]
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self.response {
+            None => Err(io::Error::new(io::ErrorKind::Other,
+                                       "Not in a readable state")),
+            Some(ref mut resp) => resp.body.read(buf),
+        }
+    }
+}
+
+impl<S> Http2Message<S> where S: CloneableStream {
+    /// A helper method that prepares the path of a request by extracting it from the given `Url`.
+    fn prepare_path(&mut self, url: Url) -> Vec<u8> {
+        let mut uri = url.serialize_path().unwrap();
+        if let Some(ref q) = url.query {
+            uri.push('?');
+            uri.push_str(&q[..]);
+        }
+        uri.into_bytes()
+    }
+
+    /// A helper method that prepares the headers that should be sent in an HTTP/2 request.
+    ///
+    /// Adapts the `Headers` into a list of octet string pairs.
+    fn prepare_headers(&mut self, mut headers: Headers) -> Vec<Http2Header> {
+        if headers.remove::<header::Connection>() {
+            warn!("The `Connection` header is not valid for an HTTP/2 connection.");
+        }
+        // HTTP/2 header names MUST be lowercase.
+        headers.iter().map(|h| {
+            (h.name().to_ascii_lowercase().into_bytes(), h.value_string().into_bytes())
+        }).collect()
+    }
+
+    /// A helper method that prepares the body for sending in an HTTP/2 request.
+    #[inline]
+    fn prepare_body(&mut self, body: Vec<u8>) -> Option<Vec<u8>> {
+        if body.len() == 0 {
+            None
+        } else {
+            Some(body)
+        }
+    }
+
+    /// Parses the response, as returned by `solicit`, into a `ResponseHead` and the full response
+    /// body.
+    ///
+    /// Returns them as a two-tuple.
+    fn parse_response(&mut self, response: ::solicit::http::Response) -> (ResponseHead, Vec<u8>) {
+        // - The status code
+        let status = response.status_code().unwrap();
+
+        // - The headers
+        let mut headers = Headers::new();
+        // Headers are moved into `hyper`'s `Headers` collection; pseudo-headers
+        // are ignored (status has already been extracted).
+        // TODO: Coalesce possible headers with the name key into a single vector.
+        //       `solicit` returns the raw original header list and lets the clients
+        //       figure out how they would like to transform it.
+        for (name, value) in response.headers.into_iter() {
+            // Unfortunately, `hyper` wants its header names to be strings...
+            // We just panic if it ends up being an invalid utf8 string, but no
+            // copies are made.
+            headers.set_raw(String::from_utf8(name).unwrap(), vec![value]);
```

Comment:
`Headers::from_raw` will do the right thing. However, `headers.set_raw()` will *override* whatever value was there before, like any setter.

The cookie formatting hack was indeed a nasty hack. We've been trying to think of a way to allow headers to indicate they output over multiple lines, but keeping that invisible from 99% of headers (since they don't need it) is difficult. I'd like to try to figure that out, such that a `Header` can return a `Encoder` that implements `Iterator`, and also be able to pass that to solicit, such that the bytes are written only once through hpack, instead of allocating a temporary vector beforehand.